### PR TITLE
docs: Add Argo CD components architecture doc

### PR DESCRIPTION
**Pull request by leoluz**
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: docs/developer-guide/architecture/components.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+to have the following benefits:
+
+- **Modularity**: Provides great level of flexibility. Components
+  interact with each other via an interface. This means that as long
+  as the interface contract is respected, a given component can be
+  replaced without requiring the rest of the system to adapt. It is
+  also possible to run the system without certain components if a
+  specific group of functionality isn't desired.
+- **Single responsibility**: Helps to determine where the different
+  types of functionality should be implementation which drives for
```
Comment (crenshaw-dev) [inline docs/developer-guide/architecture/components.md:14]:
```suggestion
  types of functionality should be implemented which drives for
```

File: docs/developer-guide/architecture/components.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+and different Kubernetes controllers. The controllers aren't
+categorized as components as they have proprietary interfaces (CRDs)
+and therefore, miss the modular nature. There are more resources
+created while installing Argo CD (ConfigMaps, Services, etc), but for
+simplicity we are covering just the ones directly related with the
+componentized architecture.
+
+## Dependencies
+
+The diagram below have represented all dependencies between the
```
Comment (crenshaw-dev) [inline docs/developer-guide/architecture/components.md:29]:
```suggestion
The diagram below has represented all dependencies between the
```

File: docs/developer-guide/architecture/components.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+### Kube API
+
+Argo CD controllers will connect to the Kubernetes API in order to run
+the reconciliation loop.
+
+### Git
+
+As a gitops tool Argo CD requires that the desired state of the
+Kubernetes resources to be provided in a Git repository.
```
Comment (crenshaw-dev) [inline docs/developer-guide/architecture/components.md:105]:
```suggestion
### Git

As a gitops tool Argo CD requires that the desired state of the
Kubernetes resources to be provided in a Git repository.

We use "git" here to stand in for an actual git repo, a Helm repo,
or an OCI artifact repo. Argo CD supports all those options.
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

**Review by crenshaw-dev — APPROVED** [review]

**Comment by cjc7373** [comment]
Is it reasonable that this doc can be merged with [operator-manual/architecture](https://argo-cd.readthedocs.io/en/latest/operator-manual/architecture/)?

**Pull request merged**